Support x86_64 in IPInt

danlliu opened the pull request:
#### 55d789ff3bf87148459a5bf12aa2b2c5ab547d56
<pre>
Support x86_64 in IPInt
<a href="https://bugs.webkit.org/show_bug.cgi?id=285022">https://bugs.webkit.org/show_bug.cgi?id=285022</a>
<a href="https://rdar.apple.com/141823156">rdar://141823156</a>

Reviewed by Yusuke Suzuki.

Previously, IPInt was not tested on x86_64 systems. This patch fixes a variety of issues
in IPInt on x86_64, mainly register conflicts during calls and fixing loading the address
of labels. This allows IPInt to run on x86_64 systems.

* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/offlineasm/x86.rb:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addArguments):
(JSC::Wasm::gprToIndex):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::FuncRefTable::setFunction):
* Tools/lldb/debug_ipint.py:

Canonical link: <a href="https://commits.webkit.org/288261@main">https://commits.webkit.org/288261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a6f722ab87af8b16315a8530345927e71fc904c6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1945 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36348 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87413 "") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84386 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2016 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9827 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/87413 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21874 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85350 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74870 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/87413 "") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1306 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29051 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32383 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75267 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29675 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88769 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81335 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9587 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6847 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72520 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9813 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70684 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71738 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15873 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14883 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/942 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12770 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9540 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15061 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103748 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9414 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25167 "Found 3 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/substring-global-atom-cache.js.lockdown, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12880 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11184 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -2633,6 +2641,17 @@ auto IPIntGenerator::endTopLevel(BlockSignature signature, const Stack& expressi
+    }
+    RELEASE_ASSERT_NOT_REACHED_UNDER_CONSTEXPR_CONTEXT();
+    return 0;
+}
```
I think this could just be `GPRInfo::toArgumentIndex(r);`

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -801,6 +801,14 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addTableCopy(unsigned dstTableI
+        }
+        RELEASE_ASSERT_NOT_REACHED_UNDER_CONSTEXPR_CONTEXT();
+        return 0;
+    };
```
Ditto on `toArgumentIndex`

kmiller68 commented on Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
```diff
@@ -126,6 +126,14 @@ void FunctionIPIntMetadataGenerator::addLEB128V128Constant(v128_t value, size_t
+        }
+        RELEASE_ASSERT_NOT_REACHED_UNDER_CONSTEXPR_CONTEXT();
+        return 0;
+    };
```
Ditto on `toArgumentIndex`

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -581,12 +626,10 @@ instructionLabel(_return_call_indirect)
     operationCall(macro() cCall4(_ipint_extern_prepare_call_indirect) end)
     btpz r1, .ipint_call_indirect_throw
+    loadp [sp], IPIntCallCallee
+    addq 16, sp
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -597,34 +640,39 @@ instructionLabel(_call_ref)
     operationCall(macro() cCall4(_ipint_extern_prepare_call_ref) end)
     btpz r1, .ipint_call_ref_throw
+    loadp [sp], IPIntCallCallee
+    addq 16, sp
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -122,7 +121,7 @@ macro popQuad(reg, scratch)
         pop scratch, reg
     elsif X86_64
-        pop reg
+        pop reg, scratch
```
I think we could make popQuad(reg) be ldr reg, sp, 16 (I would probably call it load64Inc in offlineasm) we wouldn't need to have a scratch. This could probably be a general opcode for all the CPUs or arm specific.

kmiller68 replied:
Alternatively, you could do `loadq reg; addq sp, 16` we already emit two instructions here for x86 anyway. It's just a bit more confusing since we'd be taking the scratch but relying on not using it on x86. So if you do this we should add a fix me for the above addressing mode.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -513,37 +542,46 @@ instructionLabel(_call)
     operationCall(macro() cCall3(_ipint_extern_prepare_call) end)
-    popQuad(sc1, t2)
+    loadp [sp], IPIntCallCallee
+    addq 16, sp
```
This could just be popQuad again if you do add the  "post-indexing" instruction.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -561,18 +599,25 @@ instructionLabel(_return_call)
     operationCall(macro() cCall3(_ipint_extern_prepare_call) end)
-    popQuad(sc1, t2)
+    loadp [sp], IPIntCallCallee
+    addq 16, sp
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -597,34 +640,39 @@ instructionLabel(_call_ref)
     operationCall(macro() cCall4(_ipint_extern_prepare_call_ref) end)
     btpz r1, .ipint_call_ref_throw
+    loadp [sp], IPIntCallCallee
+    addq 16, sp
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
```diff
@@ -126,6 +126,9 @@ void FunctionIPIntMetadataGenerator::addLEB128V128Constant(v128_t value, size_t
 {
+    auto gprToIndex = [&](GPRReg r) -> unsigned {
+        return GPRInfo::toArgumentIndex(r);
+    };
```
I think just using `GPRInfo::toArgumentIndex` instead of having `gprToIndex` lambda is better.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
```diff
@@ -126,6 +126,9 @@ void FunctionIPIntMetadataGenerator::addLEB128V128Constant(v128_t value, size_t
+    auto gprToIndex = [&](GPRReg r) -> unsigned {
+        return GPRInfo::toArgumentIndex(r);
+    };
     auto fprToIndex = [&](FPRReg r) -> unsigned {
```
Ditto. We should just directly use `FPRInfo::toArgumentRegister`.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -801,6 +801,9 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addTableCopy(unsigned dstTableI
 {
+    auto gprToIndex = [&](GPRReg r) -> unsigned {
+        return GPRInfo::toArgumentIndex(r);
+    };
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -801,6 +801,9 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addTableCopy(unsigned dstTableI
+    auto gprToIndex = [&](GPRReg r) -> unsigned {
+        return GPRInfo::toArgumentIndex(r);
+    };
     auto fprToIndex = [&](FPRReg r) -> unsigned {
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -2633,6 +2640,12 @@ auto IPIntGenerator::endTopLevel(BlockSignature signature, const Stack& expressi
+{
+    return GPRInfo::toArgumentIndex(r);
+}
+
```
Ditto. Just use `GPRInfo::toArgumentIndex` directly.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -2633,6 +2640,12 @@ auto IPIntGenerator::endTopLevel(BlockSignature signature, const Stack& expressi
+    return GPRInfo::toArgumentIndex(r);
+}
+
 static constexpr unsigned fprToIndex(FPRReg r)
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -365,9 +372,12 @@ instructionLabel(_rethrow)
+    popQuad(a2)
     bieq a2, ValueNull, .throw_null_ref
 
+if X86_64
+    loadp UnboxedWasmCalleeStackSlot[cfr], ws0
+end
```
Why do we need to do this only for x64?

danlliu replied:
Since x64 doesn't have as many registers, ws0 conflicts with t0 and r0, so we have to reload the callee in these situations. ARM doesn't have that register conflict, so it isn't necessary there.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -333,6 +334,9 @@ end
     nextIPIntInstruction()
 
 instructionLabel(_throw)
+if X86_64
+    loadp UnboxedWasmCalleeStackSlot[cfr], ws0
+end
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -349,6 +353,9 @@ instructionLabel(_throw)
     jumpToException()
 
 instructionLabel(_rethrow)
+if X86_64
+    loadp UnboxedWasmCalleeStackSlot[cfr], ws0
+end
```
Ditto.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -627,12 +649,20 @@ macro ipintCatchCommon()
     addq IPIntCalleeSaveSpaceStackAligned, t0
     subp cfr, t0, sp
+
+if X86_64
+    loadp UnboxedWasmCalleeStackSlot[cfr], ws0
+end
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/GPRInfo.h:
```diff
@@ -544,7 +544,7 @@ class GPRInfo {
     }
 
-    static unsigned toArgumentIndex(GPRReg reg)
+    static constexpr unsigned toArgumentIndex(GPRReg reg)
```
Do we need this change?

danlliu committed:
Support x86_64 in IPInt
https://bugs.webkit.org/show_bug.cgi?id=285022
rdar://141823156

Reviewed by Yusuke Suzuki.

Previously, IPInt was not tested on x86_64 systems. This patch fixes a variety of issues
in IPInt on x86_64, mainly register conflicts during calls and fixing loading the address
of labels. This allows IPInt to run on x86_64 systems.

* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/offlineasm/x86.rb:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addArguments):
(JSC::Wasm::gprToIndex):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmTable.cpp:
(JSC::Wasm::FuncRefTable::setFunction):
* Tools/lldb/debug_ipint.py:

Canonical link: https://commits.webkit.org/288261@main
